In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r"C:\Projects\Online+Chess+Games\chess_games.csv")
df

game_id  rated  turns victory_status winner time_increment  \
0            1  False     13    Out of Time  White           15+2   
1            2   True     16         Resign  Black           5+10   
2            3   True     61           Mate  White           5+10   
3            4   True     61           Mate  White           20+0   
4            5   True     95           Mate  White           30+3   
...        ...    ...    ...            ...    ...            ...   
20053    20054   True     24         Resign  White          10+10   
20054    20055   True     82           Mate  Black           10+0   
20055    20056   True     35           Mate  White           10+0   
20056    20057   True    109         Resign  White           10+0   
20057    20058   True     78           Mate  Black           10+0   

            white_id  white_rating            black_id  black_rating  \
0           bourgris          1500                a-00          1191   
1               a-00          1322           skinnerua          1261   
2             ischia          1496                a-00          1500   
3      daniamurashov          1439        adivanov2009          1454   
4          nik221107          1523        adivanov2009          1469   
...              ...           ...                 ...           ...   
20053        belcolt          1691            jamboger          1220   
20054       jamboger          1233  farrukhasomiddinov          1196   
20055       jamboger          1219        schaaksmurf3          1286   
20056   marcodisogno          1360            jamboger          1227   
20057       jamboger          1235               ffbob          1339   

                                                   moves opening_code  \
0      d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...          D10   
1      d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...          B00   
2      e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...          C20   
3      d4 d5 Nf3 Bf5 Nc3 Nf6 Bf4 Ng4 e3 Nc6 Be2 Qd7 O...          D02   
4      e4 e5 Nf3 d6 d4 Nc6 d5 Nb4 a3 Na6 Nc3 Be7 b4 N...          C41   
...                                                  ...          ...   
20053  d4 f5 e3 e6 Nf3 Nf6 Nc3 b6 Be2 Bb7 O-O Be7 Ne5...          A80   
20054  d4 d6 Bf4 e5 Bg3 Nf6 e3 exd4 exd4 d5 c3 Bd6 Bd...          A41   
20055  d4 d5 Bf4 Nc6 e3 Nf6 c3 e6 Nf3 Be7 Bd3 O-O Nbd...          D00   
20056  e4 d6 d4 Nf6 e5 dxe5 dxe5 Qxd1+ Kxd1 Nd5 c4 Nb...          B07   
20057  d4 d5 Bf4 Na6 e3 e6 c3 Nf6 Nf3 Bd7 Nbd2 b5 Bd3...          D00   

       opening_moves                        opening_fullname  \
0                  5        Slav Defense: Exchange Variation   
1                  4  Nimzowitsch Defense: Kennedy Variation   
2                  3   King's Pawn Game: Leonardis Variation   
3                  3  Queen's Pawn Game: Zukertort Variation   
4                  5                        Philidor Defense   
...              ...                                     ...   
20053              2                           Dutch Defense   
20054              2                            Queen's Pawn   
20055              3         Queen's Pawn Game: Mason Attack   
20056              4                            Pirc Defense   
20057              3         Queen's Pawn Game: Mason Attack   

         opening_shortname opening_response    opening_variation  
0             Slav Defense              NaN   Exchange Variation  
1      Nimzowitsch Defense              NaN    Kennedy Variation  
2         King's Pawn Game              NaN  Leonardis Variation  
3        Queen's Pawn Game              NaN  Zukertort Variation  
4         Philidor Defense              NaN                  NaN  
...                    ...              ...                  ...  
20053        Dutch Defense              NaN                  NaN  
20054         Queen's Pawn              NaN                  NaN  
20055    Queen's Pawn Game              NaN         Mason Attack

<center> Question 1 </center>
We want to determine the percentage of games won by white, black and games that ended in a draw

In [3]:
#Use the info method to determine the amount of entries in the dataframe, this shows the total number of games recorded
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   game_id            20058 non-null  int64 
 1   rated              20058 non-null  bool  
 2   turns              20058 non-null  int64 
 3   victory_status     20058 non-null  object
 4   winner             20058 non-null  object
 5   time_increment     20058 non-null  object
 6   white_id           20058 non-null  object
 7   white_rating       20058 non-null  int64 
 8   black_id           20058 non-null  object
 9   black_rating       20058 non-null  int64 
 10  moves              20058 non-null  object
 11  opening_code       20058 non-null  object
 12  opening_moves      20058 non-null  int64 
 13  opening_fullname   20058 non-null  object
 14  opening_shortname  20058 non-null  object
 15  opening_response   1207 non-null   object
 16  opening_variation  14398 non-null  objec

In [4]:
#Percentage of Games Won by both chess colours and by Draw
round(df.winner.value_counts(normalize=True)*100, 2)

winner
White    49.86
Black    45.40
Draw      4.74
Name: proportion, dtype: float64

<p> The percentage of games won by White is approximately 49.86%, the percentage of games won by Black is approximately 
45.4% & the percentage of games that ended in a draw was approximately 4.74%. </p>

<center> Question 2 </center>
We want to determine the most frequently used opening move in games where black and white won, as well as when the game ended in a draw

In [5]:
#Frequently used opening moves where Black won
q2 = df[['winner','opening_fullname']]
q2_black = q2[q2['winner'] == 'Black']
q2_black.value_counts()


winner  opening_fullname                            
Black   Van't Kruijs Opening                            226
        Sicilian Defense                                194
        Sicilian Defense: Bowdler Attack                164
        Scandinavian Defense                            123
        French Defense: Knight Variation                121
                                                       ... 
        Portuguese Opening: Portuguese Gambit             1
        Ponziani Opening: Romanishin Variation            1
        Polish Opening: Tartakower Gambit                 1
        Polish Opening: Schiffler-Sokolsky Variation      1
        Zukertort Opening: Wade Defense                   1
Name: count, Length: 1145, dtype: int64

In [6]:
#Frequently used opening moves where White won
q2 = df[['winner','opening_fullname']]
q2_white = q2[q2['winner'] == 'White']
q2_white.value_counts()

winner  opening_fullname                                                  
White   Scandinavian Defense: Mieses-Kotroc Variation                         164
        Sicilian Defense                                                      149
        Scotch Game                                                           145
        French Defense: Knight Variation                                      135
        Philidor Defense #3                                                   127
                                                                             ... 
        Scandinavian Defense: Kiel Variation                                    1
        Scandinavian Defense: Classical Variation                               1
        Pirc Defense: Classical Variation |  Quiet System                       1
        Pirc Defense: Classical Variation |  Quiet System |  Chigorin Line      1
        Zukertort Opening: Vos Gambit                                           1
Name: count, Length: 11

In [7]:
#Frequently used opening moves where the game ended in a Draw
q2 = df[['winner','opening_fullname']]
q2_draw = q2[q2['winner'] == 'Draw']
q2_draw.value_counts()

winner  opening_fullname                                               
Draw    Van't Kruijs Opening                                               16
        French Defense: Knight Variation                                   15
        Sicilian Defense                                                   15
        Queen's Pawn Game: Mason Attack                                    13
        Sicilian Defense: Bowdler Attack                                   13
                                                                           ..
        Gruenfeld Defense: Three Knights Variation |  Burille Variation     1
        Queen's Pawn Game: Colle System                                     1
        Queen's Pawn Game: Franco-Sicilian Defense                          1
        Queen's Pawn Game: Krause Variation                                 1
        Zukertort Opening: Vos Gambit                                       1
Name: count, Length: 413, dtype: int64

As we can see the most frequently used opening move where Black won and where the game ended in a draw, was the
<i>Van't Kruijs Opening</i>. For games where White won, the most frequently used opening move was the 
<i>Scandinavian Defense: Mieses-Kotroc Variation</i>

<center> Question 3 </center>
We want to find the percentage of games that were won by the player with the highest rating. Recall the conditon to be consider a high rating player is to achieve a rating of 1250 or greater.

In [33]:
q3 = df[['white_id','white_rating']]
q3

white_id  white_rating
0           bourgris          1500
1               a-00          1322
2             ischia          1496
3      daniamurashov          1439
4          nik221107          1523
...              ...           ...
20053        belcolt          1691
20054       jamboger          1233
20055       jamboger          1219
20056   marcodisogno          1360
20057       jamboger          1235

[20058 rows x 2 columns]

In [34]:
q3_white = q3[q3['white_rating'] == q3['white_rating'].max()]
q3_white

white_id  white_rating
13009  justicebot          2700

In [30]:
q3a = df[['black_id','black_rating']]
q3a

black_id  black_rating
0                    a-00          1191
1               skinnerua          1261
2                    a-00          1500
3            adivanov2009          1454
4            adivanov2009          1469
...                   ...           ...
20053            jamboger          1220
20054  farrukhasomiddinov          1196
20055        schaaksmurf3          1286
20056            jamboger          1227
20057               ffbob          1339

[20058 rows x 2 columns]

In [35]:
q3_black = q3a[q3a['black_rating'] == q3a['black_rating'].max()]
q3_black

black_id  black_rating
4473  justicebot          2723

<p>From extracting the user with the highest ratings with both colors is the user <i>justicebot</i>. Next we determine the percentage of games won by the user </p>

In [44]:
q3_final = df[['winner', 'white_id', 'black_id']]

#We want to show the times justicebot has
q3_final[(df['white_id'] == 'justicebot') & (df['winner'] == 'White') | (df['black_id'] == 'justicebot') & (df['winner'] == 'Black')]

winner    white_id       black_id
13009  White  justicebot  youredeadmeat

<p>So the user <i>justicebot</i> has only won one game out of the games recorded. </p>

In [54]:
x = round((1/20058)*100, 5)
print( x)

0.00499


<p>The user with the highest rating has only won 0.00499% of the games recorded. </p>

<center> Question 4 </center>
<p>We want to find the user that won the most games and the percentage of games where they were a high rating player</p>
<p>The website, Maven, did not specifiy the condition to be considered a high rating player, so for this case
We assume that a high rating player has a rating of 1250 or more</p>

In [8]:
# Set up dataframe to find the user that won the most games
q4 = df.iloc[:, [4,6,7,8,9]]
q4

winner       white_id  white_rating            black_id  black_rating
0      White       bourgris          1500                a-00          1191
1      Black           a-00          1322           skinnerua          1261
2      White         ischia          1496                a-00          1500
3      White  daniamurashov          1439        adivanov2009          1454
4      White      nik221107          1523        adivanov2009          1469
...      ...            ...           ...                 ...           ...
20053  White        belcolt          1691            jamboger          1220
20054  Black       jamboger          1233  farrukhasomiddinov          1196
20055  White       jamboger          1219        schaaksmurf3          1286
20056  White   marcodisogno          1360            jamboger          1227
20057  Black       jamboger          1235               ffbob          1339

[20058 rows x 5 columns]

In [9]:
#First we determine who won the most games as White
y = q4[q4['winner'] == 'White']
y['white_id'].value_counts()

white_id
taranga               34
ssf7                  29
hassan1365416         28
a_p_t_e_m_u_u         25
vladimir-kramnik-1    22
                      ..
angelkiss              1
morenobh               1
szupfer                1
andradedavi            1
marcodisogno           1
Name: count, Length: 5437, dtype: int64

In [10]:
#Now we need to check who won the most games as Black
z = q4[q4['winner'] == 'Black']
z['black_id'].value_counts()

black_id
taranga               38
ducksandcats          29
vladimir-kramnik-1    28
chesscarl             27
docboss               25
                      ..
serjjant               1
thesatantango          1
jrodneystar            1
dimehola               1
ffbob                  1
Name: count, Length: 4943, dtype: int64

<p>So as we see, the user by the name, <i>taranga</i>, won the most games overall</p>

In [11]:
#Now that we see the user that won the most games, we can now deterine the perentage of those games where they are the highest rated player

y_1 = df[(df['winner'] == 'White') & (df['white_id'] == 'taranga') & (df['white_rating'] >= 1250)]

#Apply the info method to determine the number of games where taranga won as white and was high rated player
y_1.info()
#We can see that there were 32 games where the conditions were met

<class 'pandas.core.frame.DataFrame'>
Index: 32 entries, 202 to 19089
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   game_id            32 non-null     int64 
 1   rated              32 non-null     bool  
 2   turns              32 non-null     int64 
 3   victory_status     32 non-null     object
 4   winner             32 non-null     object
 5   time_increment     32 non-null     object
 6   white_id           32 non-null     object
 7   white_rating       32 non-null     int64 
 8   black_id           32 non-null     object
 9   black_rating       32 non-null     int64 
 10  moves              32 non-null     object
 11  opening_code       32 non-null     object
 12  opening_moves      32 non-null     int64 
 13  opening_fullname   32 non-null     object
 14  opening_shortname  32 non-null     object
 15  opening_response   0 non-null      object
 16  opening_variation  25 non-null     object
dtyp

In [12]:
#We apply the same idea for when taranga won on Black and was a high rated player
y_2 = df[(df['winner'] == 'Black') & (df['black_id'] == 'taranga') & (df['black_rating'] >= 1250)]

y_2.info()
#We can see that there were 37 games where the conditions were met

<class 'pandas.core.frame.DataFrame'>
Index: 37 entries, 193 to 19090
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   game_id            37 non-null     int64 
 1   rated              37 non-null     bool  
 2   turns              37 non-null     int64 
 3   victory_status     37 non-null     object
 4   winner             37 non-null     object
 5   time_increment     37 non-null     object
 6   white_id           37 non-null     object
 7   white_rating       37 non-null     int64 
 8   black_id           37 non-null     object
 9   black_rating       37 non-null     int64 
 10  moves              37 non-null     object
 11  opening_code       37 non-null     object
 12  opening_moves      37 non-null     int64 
 13  opening_fullname   37 non-null     object
 14  opening_shortname  37 non-null     object
 15  opening_response   5 non-null      object
 16  opening_variation  21 non-null     object
dtyp

In [56]:
# Store the percentage of games won in the variable h
h = round((37+32)/20058*100,5)
print(h)

0.344


<p>The user, <i>taranga</i>, who won the most games, won approximately 0.344% of the games recorded.</p>